docs/tutorials/intro-to-rust.md

```diff
@@ -0,0 +1,418 @@
+#[near_bindgen]
+#[derive(Default, BorshDeserialize, BorshSerialize)]
+#[derive(Serialize, Deserialize)]
+pub struct Counter {
+    val: i8, // i8 is signed. unsigned integers are also available: u8, u16, u32, u64, u128
+}
+
+#[near_bindgen]
+impl Counter {
+    // returns 8-bit signed integer, must match the type from our struct's 'val' defined above
+    pub fn get_num(&self) -> i8 {
+        return self.val;
+    }
+    
+    // increment the counter
+    pub fn increment(&mut self) {
+        // note: adding one like this is an easy way to accidentally overflow
+        // real smart contracts will want to have safety checks
+        self.val = self.val + 1;
+        let _log_message = format!("Increased number to {}", self.val);
+        env::log(_log_message.as_bytes());
+        after_counter_change();
+    }
+
+    // decrement (subtract from) the counter
+    pub fn decrement(&mut self) {
+        // note: subtracting one like this is an easy way to accidentally overflow
+        // real smart contracts will want to have safety checks
+        self.val = self.val - 1;
+        let _log_message = format!("Decreased number to {}", self.val);
+        env::log(_log_message.as_bytes());
+        after_counter_change();
+    }
+}
+
+// unlike the struct's functions above, this function cannot use attributes #[derive(…)] or #[near_bindgen]
+// any attempts will throw helpful warnings upon 'cargo build'
+// while this function cannot be invoked directly on the blockchain, it can be called from an invoked function
+pub fn after_counter_change() {
+    // show helpful warning that i8 (8-bit signed integer) will overflow above 127 or below -128
+    env::log("Make sure you don't overflow, my friend.".as_bytes());
+}
+
+/*
+ * the rest of this file sets up unit tests
+ * to run these, the command will be:
+ * cargo test --package rust-counter-tutorial -- --nocapture
+ * Note: 'rust-counter-tutorial' comes from cargo.toml's 'name' key
+ */
+
+// use the attribute below for unit tests
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use near_bindgen::MockedBlockchain;
+    use near_bindgen::{testing_env, VMContext};
+
+    // part of writing unit tests is setting up a mock context
+    // in this example, this is only needed for env::log in the contract
+    fn get_context(input: Vec<u8>, is_view: bool) -> VMContext {
+        VMContext {
+            current_account_id: "alice_near".to_string(),
+            signer_account_id: "robert_near".to_string(),
+            signer_account_pk: vec![0, 1, 2],
+            predecessor_account_id: "jane_near".to_string(),
+            input,
+            block_index: 0,
+            block_timestamp: 0,
+            account_balance: 0,
+            account_locked_balance: 0,
+            storage_usage: 0,
+            attached_deposit: 0,
+            prepaid_gas: 10u64.pow(18),
+            random_seed: vec![0, 1, 2],
+            is_view,
+            output_data_receivers: vec![],
```

Comment:
@nearmax We should consider having some API for `VMContext` (builder pattern or at least Default implementation) to avoid breakage just due to the exhaustiveness of this struct initialization.